readium/lcp/src/main/java/org/readium/r2/lcp/LcpPublicationRetriever.kt

```diff
@@ -0,0 +1,214 @@
+                val container = createLicenseContainer(file, mediaType)
+                container.write(license)
+            } catch (e: Exception) {
+                tryOrLog { file.delete() }
+                listenersForId.forEach {
+                    it.onAcquisitionFailed(lcpRequestId, LcpException.wrap(e))
+                }
+                return
+            }
+
+            val acquiredPublication = LcpService.AcquiredPublication(
+                localFile = file,
+                suggestedFilename = "${license.id}.${formatRegistry.fileExtension(mediaType) ?: "epub"}",
+                mediaType = mediaType,
+                licenseDocument = license
+            )
+
+            listenersForId.forEach {
+                it.onAcquisitionCompleted(lcpRequestId, acquiredPublication)
+            }
+            listeners.remove(lcpRequestId)
+        }
+
+        override fun onDownloadProgressed(
+            requestId: DownloadManager.RequestId,
+            downloaded: Long,
+            expected: Long?
+        ) {
+            val lcpRequestId = RequestId(requestId.value)
+            val listenersForId = listeners[lcpRequestId].orEmpty()
+
+            listenersForId.forEach {
+                it.onAcquisitionProgressed(
+                    lcpRequestId,
+                    downloaded,
+                    expected
+                )
+            }
+        }
+
+        override fun onDownloadFailed(
+            requestId: DownloadManager.RequestId,
+            error: DownloadManager.Error
+        ) {
+            val lcpRequestId = RequestId(requestId.value)
+            val listenersForId = listeners[lcpRequestId].orEmpty()
+
+            listenersForId.forEach {
+                it.onAcquisitionFailed(
+                    lcpRequestId,
+                    LcpException.Network(Exception(error.message))
+                )
+            }
+
+            listeners.remove(lcpRequestId)
+        }
+
+        override fun onDownloadCancelled(requestId: DownloadManager.RequestId) {
+            val lcpRequestId = RequestId(requestId.value)
+            val listenersForId = listeners[lcpRequestId].orEmpty()
+            listenersForId.forEach {
+                it.onAcquisitionCancelled(lcpRequestId)
+            }
+            listeners.remove(lcpRequestId)
+        }
+    }
+
+    private val formatRegistry: FormatRegistry =
+        FormatRegistry()
+
+    private val downloadsRepository: LcpDownloadsRepository =
+        LcpDownloadsRepository(context)
+
+    private val downloadListener: DownloadManager.Listener =
+        DownloadListener()
+
+    private val listeners: MutableMap<RequestId, MutableList<Listener>> =
+        mutableMapOf()
+
+    public fun register(
+        requestId: RequestId,
+        listener: Listener
+    ) {
+        listeners.getOrPut(requestId) { mutableListOf() }.add(listener)
+        downloadManager.register(DownloadManager.RequestId(requestId.value), downloadListener)
```

Comment:
I think we will get duplicated calls if we have several listeners, as we register to the download manager again.
```suggestion
        if (listeners[requestId].size() == 1) {
            downloadManager.register(DownloadManager.RequestId(requestId.value), downloadListener)
        }
```